rackhd_BIST/test_utils.py

```diff
@@ -0,0 +1,208 @@
+    logger.addHandler(logfile)
+    logger.addHandler(console)
+    return logger
+
+def tool_version_compare(version_a, version_b):
+    """
+    Compare for tool version:
+    :param version_a: version to be compared
+    :param version_b: version to be compared
+    :return  1:  version_a is larger than version_b
+            0:  version_a equals version b
+            -1: version-a is smaller than version b
+    """
+    version_a_bits = version_a.split(".")
+    version_b_bits = version_b.split(".")
+    if version_a_bits > version_b_bits:
+        return 1
+    elif version_a_bits < version_b_bits:
+        return -1
+    else:
+        return 0
+
+def get_tool_version(cmd, redirect=False):
+    """
+    Use given command to get tool version
+    :param cmd: command to get tool version
+    :param redirect: flag decide if we should redirect STDERR to STDOUT
+    :return: tool version string
+    """
+    result = robust_check_output(cmd=cmd, shell=False, redirect=redirect)
+    pattern = re.compile(r'\d\.(\d\.)*\d')
+    if result["exit_code"] == 0:
+        version = pattern.search(result["message"])
+        if version:
+            result["message"] = version.group(0) ## refine output message
+        else:
+            result["exit_code"] = -1
+            result["message"] = "Failed to get version with command [{}]".format(",".join(cmd))
+    return result
+
+CONFIGURATION = get_configurations()
+
+class Logger(object):
+    """
+    RackHD BIST specified logging class
+    """
+    def __init__(self):
+        self.logger = initiate_logger(
+            "rackhd_bist_log",
+            CONFIGURATION.get('logPath')
+        )
+
+    def record_log_message(self, description, details, level):
+        """
+        Record RackHD BIST log message in a unified format
+        :param description: short log description for a test, required
+        :param details: detailed information for a test, optional
+        :param level: logging level, one of ["debug", "info", "warning", "error" ]
+        """
+        assert level in ["info", "warning", "error", "debug"], "Logging level is incorrect"
+        log_message = '[message]: {}  '.format(description)
+        if details:
+            log_message = log_message + '[details]: {}'.format(details)
+        logger_method = getattr(self.logger, level)
+        logger_method(log_message)
+
+    def record_command_result(self, description, status, level):
+        """
+        Logging according command output
+        :param status: return of robust_check_output, robust_open_file or robust_load_json_file,
+            an example:
+            {"exit_code": 0, "message": "check_call command succeeded"}
+        :param description: short log description for a test
+        :param level: logging level if test failed
+        """
+        details = status["message"].strip("\n")
+        if status["exit_code"] == 0:
+            level = 'debug'
+            description = description + " succeeded"
```

Comment:
the input `description` may be anything, how you make sure after appending the word `succeeded` is still a valid sentence? The `success` or `failure` is an important status, so it worth to specially output it just like the level.